src/tools/w3dviewer/scenelightdialog.cpp

```diff
@@ -0,0 +1,312 @@
+        }
+
+        SetDlgItemFloat(m_hWnd, IDC_START, start);
+        SetDlgItemFloat(m_hWnd, IDC_END, end);
+        SetDlgItemFloat(m_hWnd, IDC_DISTANCE, distance);
+
+        m_distanceSpin.SetRange(0, 16960);
+        m_distanceSpin.SetPos(distance * 100.0f);
+        m_startSpin.SetRange(0, 16960);
+        m_startSpin.SetPos(start * 100.0f);
+        m_endSpin.SetRange(0, 16960);
+        m_endSpin.SetPos(end * 100.0f);
+        m_intensitySlider.SetRange(0, 100);
+        m_intensitySlider.SetPos(intensity * 100.0f);
+
+        m_farAttenStart = start;
+        m_farAttenEnd = end;
+        m_distance = distance;
+        m_intensity = intensity;
+        m_attenuation = attenuation;
+    }
+
+    SendDlgItemMessage(IDC_DIFFUSE, BM_SETCHECK, TRUE, 0);
+    bool enable = SendDlgItemMessage(IDC_ATTENUATION, BM_GETCHECK) == 1;
+    GetDlgItem(IDC_START)->EnableWindow(enable);
+    GetDlgItem(IDC_STARTSPIN)->EnableWindow(enable);
+    GetDlgItem(IDC_END)->EnableWindow(enable);
+    GetDlgItem(IDC_ENDSPIN)->EnableWindow(enable);
+
+    return TRUE;
+}
+
+void CSceneLightDialog::OnCancel()
+{
+    CW3DViewDoc *doc = GetCurrentDocument();
+
+    if (doc != nullptr && doc->m_scene != nullptr) {
+        doc->m_light->Set_Diffuse(Vector3(m_diffuseRed / 100.0f, m_diffuseGreen / 100.0f, m_diffuseBlue / 100.0f));
+        doc->m_light->Set_Specular(Vector3(m_specularRed / 100.0f, m_specularGreen / 100.0f, m_specularBlue / 100.0f));
+        doc->m_light->Set_Intensity(m_intensity);
+        doc->m_light->Set_Far_Attenuation_Range(m_farAttenStart, m_farAttenEnd);
+        doc->m_light->Set_Flag(LightClass::FAR_ATTENUATION, m_attenuation == 1);
+        SetDistance(m_distance);
+    }
+
+    CDialog::OnCancel();
+}
+
+void CSceneLightDialog::OnHScroll(UINT nSBCode, UINT nPos, CScrollBar *pScrollBar)
+{
+    if (pScrollBar == GetDlgItem(IDC_INTENSITYSLIDER)) {
+        GetCurrentDocument()->m_light->Set_Intensity(m_intensitySlider.GetPos() / 100.0f);
+    } else {
+        if (SendDlgItemMessage(IDC_GREYSCALE, BM_GETCHECK)) {
+            HWND handle;
+
+            if (pScrollBar == GetDlgItem(IDC_REDSLIDER)) {
+                handle = m_redSlider.m_hWnd;
+            } else if (pScrollBar == GetDlgItem(IDC_GREENSLIDER)) {
+                handle = m_greenSlider.m_hWnd;
+            } else {
+                handle = m_blueSlider.m_hWnd;
+            }
+
+            int value = ::SendMessage(handle, TBM_GETPOS, 0, 0);
+            m_redSlider.SetPos(value);
+            m_greenSlider.SetPos(value);
+            m_blueSlider.SetPos(value);
+        }
+
+        CW3DViewDoc *doc = GetCurrentDocument();
+
+        if (doc != nullptr && doc->m_scene != nullptr) {
+            Vector3 color;
+            color.X = m_redSlider.GetPos() / 100.0f;
+            color.Y = m_greenSlider.GetPos() / 100.0f;
+            color.Z = m_blueSlider.GetPos() / 100.0f;
+
+            if ((m_flags & 2) != 0) {
```

Comment:
Can we use enum bit flag for this number?

Reply:
fixed.